### DISCUSSION TITLE
Host behind reverse proxy on path like localhost/photos

### DISCUSSION THREAD

**Discussion by LouDou** [photoview/photoview#392] [Q&A and Troubleshooting]
**Describe the bug**

I've been trying to get this app set up using docker-compose and behind an nginx reverse proxy.
My nginx instance proxies on to many apps on my server, each of them on their own path - because I'm hosting on local home network and I cannot use subdomains (there's no domain available).

My intention is to be able to host photoview under http:/[host]/photos URL; and testing locally using http://localhost/photos

However, when I've tried to configure this, the photoview app is still trying to load JS/CSS assets from the http://[host]/ root path, and there's some other strangeness going on that requesting any resource under http://[host]/photos/file.js returns the HTML from the index.html file.

**To Reproduce**

The steps required to get this going are not clear. I've tried a few different things in various permutations and none work.
I'm pasting here only the relevant parts of the config that I've tried relating to URLs and proxying; I have other docker-compose apps already working behind the proxy, and I have also got photoview to run correctly locally without the proxy.

1. photos/docker-compose.yml - using the example
-> https://github.com/photoview/photoview/blob/master/docker-compose.example.yml

2. photos/docker-compose.yml - using locally built image from git source
```
services:
  photoview:
    build:
      context: https://github.com/photoview/photoview.git
      args:
        - TARGETPLATFORM=linux/amd64
<snip>
```

3. photos/docker-compose.yml - with the `UI_PUBLIC_URL` arg:
```
  photoview:
    build:
      context: https://github.com/photoview/photoview.git
      args:
        - UI_PUBLIC_URL=/photos/
        - TARGETPLATFORM=linux/amd64
<snip>
```

4. nginx location - attempt 1
```
<snip>
location /photos {
  proxy_set_header        Host $host;
  proxy_set_header        X-Real-IP $remote_addr;
  proxy_set_header        X-Forwarded-For $proxy_add_x_forwarded_for;
  proxy_set_header        X-Forwarded-Proto $scheme;

  proxy_pass http://photoview;

  proxy_read_timeout      600s;
  proxy_send_timeout      600s;
}
<snip>
```

5. nginx location - attempt 2
```
<snip>
location /photos {
  proxy_set_header        Host $host;
  proxy_set_header        X-Real-IP $remote_addr;
  proxy_set_header        X-Forwarded-For $proxy_add_x_forwarded_for;
  proxy_set_header        X-Forwarded-Proto $scheme;

  proxy_pass http://photoview/;

  proxy_read_timeout      600s;
  proxy_send_timeout      600s;
}
<snip>
```

6. nginx location - attempt 3
```
<snip>
location /photos {
  proxy_set_header        Host $host;
  proxy_set_header        X-Real-IP $remote_addr;
  proxy_set_header        X-Forwarded-For $proxy_add_x_forwarded_for;
  proxy_set_header        X-Forwarded-Proto $scheme;

  proxy_pass http://photoview/photos/;

  proxy_read_timeout      600s;
  proxy_send_timeout      600s;
}
<snip>
```

**Expected behavior**

I'm expecting the app to load and work at http://[host]/photos

**Your environment:**
Feel free to remove and add what you think might be useful.

 - Are you using Docker? yes
 - Database mariadb 10.5
 - Browser chrome
 - Version latest / master

---

If there's any further information you require to debug this, do let me know.

Regards,
Doug.

---

Edit 1 : url markup

**Comment by viktorstrate** [comment]
I don't know if Nginx can do it but Traefik has an option called [StripPrefix](https://doc.traefik.io/traefik/middlewares/stripprefix/) that can remove the prefix before sending the request to Photoview. That way Photoview never even sees the prefix.

If your reverse proxy cannot be configured in that way, I guess it would also be possible to manually build the Docker image with a hard-coded path prefix, by setting `UI_PUBLIC_URL=/photos` at built time.

**Reply by LouDou** [reply]
Unfortunately I'm not 100% sure this can be solved with proxy config alone, although I have tried and will try again.
What I have noticed is that even when built with `UI_PUBLIC_URL=/photos` the index.html file in the image still has absolute paths for CSS and JS asset sources pointing to root "/".
I will try to make a local build from git to make sure I understand this option and see if I can get it working outside of the whole docker environment.

**Comment by LouDou** [comment]
Testing out the docker build args:

## Attempt 1 - vanilla config:
```
$ docker build --build-arg TARGETPLATFORM=linux/amd64 .
$ docker run --entrypoint /bin/bash --rm -it $IMAGEID
# cat /ui/index.html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link rel="manifest" href="/manifest.json" />
    <link rel="stylesheet" href="/index.css" />

    <!-- Apple touch devices -->
    <link rel="apple-touch-icon" href="/assets/apple-touch-icon.png" />
    <meta name="apple-mobile-web-app-title" content="Photoview" />
    <meta name="apple-mobile-web-app-capable" content="yes" />
    <meta name="apple-mobile-web-app-status-bar-style" content="white" />
  </head>
  <body>
    <div id="root"></div>
    <script type="module" src="/index.js"></script>
  </body>
</html>

# head /ui/index.js
import{a as Mt,b as Dt}from"/chunk-AWDC6VIJ.js"; ...
```



## Attempt 2 - using UI_PUBLIC_URL:

```
$ docker build --build-arg TARGETPLATFORM=linux/amd64 --build-arg UI_PUBLIC_URL=/photos/ .
$ docker run --entrypoint /bin/bash --rm -it $IMAGEID

# cat /ui/index.html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link rel="manifest" href="/manifest.json" />
    <link rel="stylesheet" href="/index.css" />

    <!-- Apple touch devices -->
    <link rel="apple-touch-icon" href="/assets/apple-touch-icon.png" />
    <meta name="apple-mobile-web-app-title" content="Photoview" />
    <meta name="apple-mobile-web-app-capable" content="yes" />
    <meta name="apple-mobile-web-app-status-bar-style" content="white" />
  </head>
  <body>
    <div id="root"></div>
    <script type="module" src="/index.js"></script>
  </body>
</html>

# head /ui/index.js
import{a as Mt,b as Dt}from"/photos/chunk-AWDC6VIJ.js"; ...
```

----


You can see that the JS module import paths have changed, but not the references to the files in the index.html - so when running behind a proxy, the app does not even load the JS.

**Reply by viktorstrate** [reply]
I see. When using `PHOTOVIEW_SERVE_UI=1` the server always expects the UI to be served at `/`.

You could fix this by serving the UI yourself, using Nginx for example, and then setting `PHOTOVIEW_SERVE_UI=0`.
Then you would set `PHOTOVIEW_API_ENDPOINT` to the path where the API server is accessed publicly and `PHOTOVIEW_UI_ENDPOINT` to the path for where the UI is served.

Maybe something like this:
```
PHOTOVIEW_SERVE_UI=0
PHOTOVIEW_API_ENDPOINT=/photos-api/
PHOTOVIEW_UI_ENDPOINT=/photos
```

You would have to use an Nginx config similar to this to serve the `index.html` file if the file doesn't exist, this is how routing is done.
```nginx
server {
    root /path/to/ui;
    location / {
        try_files $uri $uri/ /index.html; # fallback to index.html if file doesn't exist
    }
}
```

---
The above only works if the api and ui are served on the same domain, eg. with different paths, which is obviously not a problem for you as you don't have a domain to begin with. This is because of this issue: #184

**Comment by LouDou** [comment]
I've been looking at all the places that need to support the path prefix for proxying - there are many.
- the configuration in server.go is missing support
```
diff --git a/api/server.go b/api/server.go
index 8c1d57d..b1f3006 100644
--- a/api/server.go
+++ b/api/server.go
@@ -111,7 +111,7 @@ func main() {

        if shouldServeUI {
                spa := routes.NewSpaHandler(utils.UIPath(), "index.html")
-               rootRouter.PathPrefix("/").Handler(spa)
+               rootRouter.PathPrefix("/PREFIX").Handler(http.StripPrefix("/PREFIX", spa))
        }

        if devMode {
```

- The API endpoint configuration is wrong/missing support
```diff --git a/api/utils/Endpoints.go b/api/utils/Endpoints.go
index 021be41..f1a3a90 100644
--- a/api/utils/Endpoints.go
+++ b/api/utils/Endpoints.go
@@ -4,19 +4,13 @@ import (
        "fmt"
        "log"
        "net/url"
-       "path"
        "strconv"
 )

 func ApiListenUrl() *url.URL {
        const defaultPort = "4001"

-       shouldServeUI := ShouldServeUI()
-
-       apiPrefix := "/"
-       if shouldServeUI {
-               apiPrefix = "/api"
-       }
+       apiPrefix := ApiEndpointUrl().String()

        var listenAddr string

@@ -45,24 +39,11 @@ func ApiListenUrl() *url.URL {
 }

 func ApiEndpointUrl() *url.URL {
-       var apiEndpointStr string
-
-       shouldServeUI := ShouldServeUI()
-       if shouldServeUI {
-               apiEndpointStr = "/"
-       } else {
-               apiEndpointStr = EnvAPIEndpoint.GetValue()
-       }
-
-       apiEndpointURL, err := url.Parse(apiEndpointStr)
+       apiEndpointURL, err := url.Parse(EnvAPIEndpoint.GetValue())
        if err != nil {
                log.Fatalf("ERROR: Environment variable %s is not a proper url (%s)", EnvAPIEndpoint.GetName(), EnvAPIEndpoint.GetValue())
        }

-       if shouldServeUI {
-               apiEndpointURL.Path = path.Join(apiEndpointURL.Path, "/api")
-       }
-
        return apiEndpointURL
 }
diff --git a/ui/src/apolloClient.ts b/ui/src/apolloClient.ts
index c239aa8..8f65c9e 100644
--- a/ui/src/apolloClient.ts
+++ b/ui/src/apolloClient.ts
@@ -17,7 +17,7 @@ import { MessageState } from './components/messages/Messages'
 import { Message } from './components/messages/SubscriptionsHook'

 export const GRAPHQL_ENDPOINT = process.env.PHOTOVIEW_API_ENDPOINT
-  ? urlJoin(process.env.PHOTOVIEW_API_ENDPOINT, '/graphql')
+  ? urlJoin(location.origin, process.env.PHOTOVIEW_API_ENDPOINT, '/graphql')
   : urlJoin(location.origin, '/api/graphql')

 const httpLink = new HttpLink({
```
But of course, redefining it like this will break the serve UI options and also possibly running the project locally, so this is still an incomplete solution.


- the entire React Router is missing support
```
No diff here, I haven't enough experience in the Router config
```

- the build process for URLs baked into the index.html and manifest is missing support
```
diff --git a/ui/src/index.html b/ui/src/index.html
index 638830f..e375f9a 100644
--- a/ui/src/index.html
+++ b/ui/src/index.html
@@ -3,17 +3,17 @@
   <head>
     <meta charset="utf-8" />
     <meta name="viewport" content="width=device-width, initial-scale=1.0" />
-    <link rel="manifest" href="/manifest.json" />
-    <link rel="stylesheet" href="/index.css" />
+    <link rel="manifest" href="/PREFIX/manifest.json" />
+    <link rel="stylesheet" href="/PREFIX/index.css" />

     <!-- Apple touch devices -->
-    <link rel="apple-touch-icon" href="/assets/apple-touch-icon.png" />
+    <link rel="apple-touch-icon" href="/PREFIX/assets/apple-touch-icon.png" />
     <meta name="apple-mobile-web-app-title" content="Photoview" />
     <meta name="apple-mobile-web-app-capable" content="yes" />
     <meta name="apple-mobile-web-app-status-bar-style" content="white" />
   </head>
   <body>
     <div id="root"></div>
-    <script type="module" src="/index.js"></script>
+    <script type="module" src="/PREFIX/index.js"></script>
   </body>
 </html>
diff --git a/ui/src/manifest.webmanifest b/ui/src/manifest.webmanifest
index 36ff4e9..e403472 100644
--- a/ui/src/manifest.webmanifest
+++ b/ui/src/manifest.webmanifest
@@ -2,21 +2,21 @@
   "name": "Photoview",
   "short_name": "Photoview",
   "description": "Photo gallery for self-hosted personal servers",
-  "start_url": "/",
+  "start_url": "/PREFIX",
   "display": "standalone",
   "theme_color": "black",
   "background_color": "white",
   "icons": [
     {
-      "src": "/assets/logo192.png",
+      "src": "/PREFIX/assets/logo192.png",
       "sizes": "192x192"
     },
     {
-      "src": "/assets/logo512.png",
+      "src": "/PREFIX/assets/logo512.png",
       "sizes": "512x512"
     },
     {
-      "src": "/assets/photoview-logo.svg",
+      "src": "/PREFIX/assets/photoview-logo.svg",
       "sizes": "150x150"
     }
   ]
```

There's probably more, but this runs so deep in the project that I'm about ready to give up. It's a shame, I'd love to be able to use this.

**Comment by Don-Swanson** [comment]
Have you tried looking into nginx rewriting? That's what most people use when trying to do what you are doing with various services.

**Comment by kcris** [comment]
I have created an issue for this problem
https://github.com/photoview/photoview/issues/503
